modules/iothub/src/iothub.c

```diff
@@ -759,8 +913,40 @@ static void IotHub_Receive(MODULE_HANDLE moduleHandle, MESSAGE_HANDLE messageHan
                             }
                             else
                             {
+                                MESSAGE_DELIVERED_CALLBACK_CONTEXT* userContextCallback = NULL;
+                                IOTHUB_CLIENT_EVENT_CONFIRMATION_CALLBACK eventConfirmationCallback = NULL;
+
+                                /*Codes_SRS_IOTHUBMODULE_99_004: [ If the message contains a property "iotHubMessageId" then callback function `receiveMessageConfirmation` and userContext is given to `IoTHubClient_SendEventAsync` as parameters ]*/
+                                /*Codes_SRS_IOTHUBMODULE_99_005: [ If the message does not contain property "iotHubMessageId" then no callback function is given to `IoTHubClient_SendEventAsync` as a parameter ]*/
+                                const char* iotHubMessageId = ConstMap_GetValue(properties, GW_IOTHUB_MESSAGE_ID);
+                                if (iotHubMessageId)
+                                {
+                                    userContextCallback = malloc(sizeof(MESSAGE_DELIVERED_CALLBACK_CONTEXT));
+                                    if (userContextCallback == NULL)
+                                    {
+                                        /*Codes_SRS_IOTHUBMODULE_99_008: [ If memory allocation fail when handling "iotHubMessageId" property, `IoTHubClient_SendEventAsync` returns without sending the message ]*/
+                                        LogError("Failed to create MESSAGE_DELIVERED_CALLBACK_CONTEXT");
+                                        IoTHubMessage_Destroy(iotHubMessage);
+                                        ConstMap_Destroy(properties);
+                                        return;
+                                    }
+
+                                    if (mallocAndStrcpy_s(&(userContextCallback->iotHubMessageId), iotHubMessageId) != 0)
+                                    {
+                                        /*Codes_SRS_IOTHUBMODULE_99_008: [ If memory allocation fail when handling "iotHubMessageId" property, `IoTHubClient_SendEventAsync` returns without sending the message ]*/
+                                        LogError("Failed to allocate/copy iotHubMessageId");
+                                        free(userContextCallback);
+                                        IoTHubMessage_Destroy(iotHubMessage);
+                                        ConstMap_Destroy(properties);
+                                        return;
+                                    }
```

Comment:
Do we want to return here (and in the section above) or make a best effort at sending the message to IoTHub?

Reply:
Good question. If malloc fails how likely it is that sending message will success? If user needs a notification about delivery, and doesn't get one, wouldn't that be a huge problem too?

Reply:
The most likely way is if the "iotHubMessageId" string gets trashed in memory and the malloc_and_Strcpy_s fails, but even then it's unlikely that the IoTHubClient_SendEventAsync will succeed as well.  

So it's pretty unlikely it worth trying to send the message. I'm OK with leaving it alone, just more curious if you had a different answer..